Google Spreadsheet is glitchy in Safari

rniwa opened the pull request:
#### 8dbebaa6d56be48eca38f560fb23754b279ff69d
<pre>
Google Spreadsheet is glitchy in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=294310">https://bugs.webkit.org/show_bug.cgi?id=294310</a>

Reviewed by Simon Fraser.

The bug was caused by WebKit firing resize event before visibilitychange event.
Defer the dispatching of resize event while the page is not visible to emulate the behavior of Chrome.

This fixes the glitch on Google spreadsheet.

* LayoutTests/fast/events/page-visibility-resize-event-order-expected.txt: Added.
* LayoutTests/fast/events/page-visibility-resize-event-order.html: Added.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged):
(WebCore::Document::runResizeSteps):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/296124@main">https://commits.webkit.org/296124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b87bcfa1766b71d6d7d5d108bd48ef35a3c30bc2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107423 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17511 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112637 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57959 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109387 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35605 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81551 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110352 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22012 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96819 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61927 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21453 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14950 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57403 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14984 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34490 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25422 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90590 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34865 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93068 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90327 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35241 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13022 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30233 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17375 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34411 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39952 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34157 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37512 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2577,8 +2577,16 @@ void Document::unregisterForVisibilityStateChangedCallbacks(VisibilityChangeClie
+    eventLoop().queueTask(TaskSource::UserInteraction, [document = Ref { *this }, pageIsVisible = pageIsVisible] {
+        document->dispatchEvent(Event::create(eventNames().visibilitychangeEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
+        if (pageIsVisible)
+            document->m_deferResizeEventForVisibilityChange = false;
```
How do we know that `runResizeSteps()` will get called after this and the resize events will eventually fire? This boolean just makes `runResizeSteps()` early return so it stops the event firing unless `runResizeSteps()` eventually gets called again. cc @smfr

rniwa replied:
Yeah, Simon pointed out the same thing. I guess we should schedule a rendering update here when either resize event is scheduled.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2577,8 +2577,20 @@ void Document::unregisterForVisibilityStateChangedCallbacks(VisibilityChangeClie
-    queueTaskToDispatchEvent(TaskSource::UserInteraction, Event::create(eventNames().visibilitychangeEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
+    if (!pageIsVisible)
+        m_deferResizeEventForVisibilityChange = true;
+    eventLoop().queueTask(TaskSource::UserInteraction, [this, protectedDocument = Ref { *this }] {
```
I'd prefer a blank line above this one.

rniwa replied:
Sure, I can add that.

rniwa committed:
Google Spreadsheet is glitchy in Safari
https://bugs.webkit.org/show_bug.cgi?id=294310

Reviewed by Simon Fraser.

The bug was caused by WebKit firing resize event before visibilitychange event.
Defer the dispatching of resize event while the page is not visible to emulate the behavior of Chrome.

This fixes the glitch on Google spreadsheet.

* LayoutTests/fast/events/page-visibility-resize-event-order-expected.txt: Added.
* LayoutTests/fast/events/page-visibility-resize-event-order.html: Added.
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityStateChanged):
(WebCore::Document::runResizeSteps):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/296124@main
